oeps/oep-0017-feature-toggles.rst

```diff
@@ -0,0 +1,927 @@
+requests the value of a feature toggle, the framework should first check if
+there's an override for the current site and return it instead.
+
+.. _edX Site Configuration: https://github.com/edx/edx-platform/blob/3290bce9932916439e3ff8244ba34dd1444e0702/openedx/core/djangoapps/site_configuration/__init__.py#L6
+
+Req 10: Least Privilege
+^^^^^^^^^^^^^^^^^^^^^^^
+
+If business-sensitive toggles are used that need to have limited access, the
+framework should be extended to support fine-grained write access to feature
+toggles. One possibility is to add a new "group access" field with each toggle
+and update the Django admin interface to enforce access.
+
+Req 11: Discoverability
+^^^^^^^^^^^^^^^^^^^^^^^
+
+The framework needs to be able to discover all waffle_utils_ classes declared
+in all installed Django apps in the system. Initially, the discoverability can
+be scoped to within each microservice, but ultimately accessed via a centralized
+tool across all distributed services.
+
+To support this, the framework can make use of the `Django App Plugin`_ design
+pattern and search for waffle_utils_ classes declared in all installed apps. This
+requires that every app that uses waffle_utils_ declares its usages in a standard
+module (i.e., config.py) or configure its location (in its apps.py module).
+
+.. _Django App Plugin: https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/plugins/README.rst
+
+Req 12: Report
+^^^^^^^^^^^^^^
+
+In order to provide a useful and informative administrative report of the
+existing feature toggles in the system, the framework needs to be able to
+present the following information for each toggle.
+
+.. list-table::
+   :header-rows: 1
+
+   * - Report data
+     - Purpose
+     - Data source
+   * - Description
+     - Brief human-readable information about its usage and context.
+     - In code, by developer
+   * - Feature Category
+     - Optional field to group interdependent toggles.
+     - In code, by developer
+   * - All Use Cases
+     - Lists one or more `Use Cases`_ to specify all expected usages of this toggle.
+     - In code, by developer
+   * - Current Use Case(s)
+     - A subset of "All Use Cases" to specify the current `Use Cases`_ of this toggle.
+     - In code, by developer; optionally editable via admin interface.
+   * - Toggle Type
+     - One of Switch, Rollout, or Group to further clarify the toggle's usage.
+     - In code, by developer
+   * - Created in Code Date
+     - Required field to specify the date the toggle was added to the codebase; to easily find all stale toggles.
+     - In code, by developer
+   * - Expiration Date
+     - Optional field to specify target date of removal; to easily find all expired toggles.
+     - In code, by developer
+   * - Current Setting(s)
+     - Summary of the current configuration and value of the feature toggle; to easily evaluate its readiness to transition or retire.
```

Comment:
@nasthagiri: The following is not a required change.  Just a note that the reporting tool should report on the total list of coded toggles + the total list of configured toggles, so we can report orphaned configurations, where a value was found, but we have no known corresponding toggle declaration.

We did some flag renaming in the past and forgot to do clean-up of old values.  Also, it is possible that someone could remove a toggle from code and forgot to later remove the config from admin for all environments.